sdk/core/azure-core-serializer-avro-apache/src/main/java/com/azure/core/serializer/avro/apache/ApacheAvroSerializer.java

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.serializer.avro.apache;
+
+import com.azure.core.serializer.AvroSerializer;
+import com.azure.core.util.logging.ClientLogger;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.Encoder;
+import org.apache.avro.io.EncoderFactory;
+import reactor.core.Exceptions;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Apache based implementation of the {@link AvroSerializer} interface.
+ */
+public class ApacheAvroSerializer implements AvroSerializer {
+    private final ClientLogger logger = new ClientLogger(ApacheAvroSerializer.class);
+
+    private final Schema.Parser parser;
+    private final DecoderFactory decoderFactory;
+    private final EncoderFactory encoderFactory;
+    private final GenericData genericData;
+
+    ApacheAvroSerializer(Schema.Parser parser, DecoderFactory decoderFactory, EncoderFactory encoderFactory,
+        GenericData genericData) {
+        this.parser = parser;
+        this.decoderFactory = decoderFactory;
+        this.encoderFactory = encoderFactory;
+        this.genericData = genericData;
+    }
+
+    @Override
+    public <T> T read(byte[] input, String schema) {
+        Objects.requireNonNull(schema, "'schema' cannot be null.");
+
+        return readInternal(input, schema, schema);
+    }
+
+    @Override
+    public <T> T read(byte[] input, String readerSchema, String writerSchema) {
+        Objects.requireNonNull(readerSchema, "'readerSchema' cannot be null.");
+        Objects.requireNonNull(writerSchema, "'writerSchema' cannot be null.");
+
+        return readInternal(input, readerSchema, writerSchema);
+    }
+
+    private <T> T readInternal(byte[] input, String readerSchema, String writerSchema) {
+        if (input == null) {
+            return null;
+        }
+
+        try {
+            return new GenericDatumReader<T>(parser.parse(writerSchema), parser.parse(readerSchema), genericData)
```

Comment:
Isn't parsing schema for every entry extremely inefficient?

Reply:
There is some inefficiencies parsing the schema every time. The other option is keeping track of all schemas that have been parsed in a map which adds memory overhead. Could look into keeping a mapping that leverages a least recently used cache.